### PR TITLE
Reduce file usage for working with PSI

### PR DESCRIPTION
**Pull request by eartser**
Implemented PSI file factory to reuse existing files while working with PSI

### CODE REVIEW

File: code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiUtil.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+fun PsiElement.getElementChildren(toIgnoreWhiteSpaces: Boolean): List<PsiElement> = if (toIgnoreWhiteSpaces) {
+    this.children.filterWhiteSpaces().toList()
+} else {
+    this.children.toList()
+}
+
+// TODO: add other symbols?
+fun String.trimCode() = this.trim(' ', '\n')
```
Comment (nbirillo) [inline code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiUtil.kt:26]: Should we also add the system line separator?

File: code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiUtil.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package org.jetbrains.research.code.submissions.clustering.util.psi
+
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.psi.PsiElement
+import com.intellij.psi.codeStyle.CodeStyleManager
+import org.jetbrains.research.code.submissions.clustering.load.context.builder.gumtree.converter.PsiToGumTreeConverter.filterWhiteSpaces
+
+val PsiElement.isLeaf: Boolean
+    get() = this.children.isEmpty()
```
Comment (nbirillo) [inline code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiUtil.kt:9]: Should we use getElementChildren here?

File: code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiFileFactory.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }).also { it.setText(text) }
+
+    private fun PsiFile.setText(text: String) {
+        val document = viewProvider.document
+            ?: error("No document for file found")
+        PsiDocumentManager.getInstance(project).doPostponedOperationsAndUnblockDocument(document)
+        WriteAction.run<Throwable> {
+            document.setText(text)
+        }
+        PsiDocumentManager.getInstance(project).commitDocument(document)
```
Comment (nbirillo) [inline code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiFileFactory.kt:49]: Can we create a variable with the manager and reuse it?

**Review by nbirillo — CHANGES_REQUESTED** [review]
It looks fine, thanks!

File: code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiUtil.kt (hunk 2 above)
Comment (eartser) [reply code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiUtil.kt:9]: I guess, this implementation is just fine, getElementChildren will cause unnecessary memory allocation and complicate the logic.

File: code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiUtil.kt (hunk 1 above)
Comment (eartser) [reply code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/psi/PsiUtil.kt]: Also added '\r' here for now, as trim only works with chars and system line separator is a string.

**Review by nbirillo — APPROVED** [review]
Great!

**Pull request merged**